File: android/src/main/kotlin/com/oneplusdream/flutter_aws_chime/AudioVideoObserver.kt
package com.oneplusdream.flutter_aws_chime

import com.amazonaws.services.chime.sdk.meetings.audiovideo.AudioVideoObserver
import com.amazonaws.services.chime.sdk.meetings.audiovideo.video.RemoteVideoSource
import com.amazonaws.services.chime.sdk.meetings.session.MeetingSessionStatus

class AudioVideoObserver(val methodChannel: MethodChannelCoordinator) : AudioVideoObserver {
    override fun onAudioSessionCancelledReconnect() {
        methodChannel.callFlutterEvent("audioSessionReconnectCancelled")
    }

    override fun onAudioSessionDropped() {
        methodChannel.callFlutterEvent("audioSessionDropped")
    }

    override fun onAudioSessionStarted(reconnecting: Boolean) {
        methodChannel.callFlutterEvent(
                "audioSessionStarted",
                mapOf("reconnecting" to reconnecting)
        )
    }

    override fun onAudioSessionStartedConnecting(reconnecting: Boolean) {
        methodChannel.callFlutterEvent(
                "audioSessionConnecting",
                mapOf("reconnecting" to reconnecting)
        )
    }

    override fun onAudioSessionStopped(sessionStatus: MeetingSessionStatus) {
        MeetingSessionManager.onSessionStopped()
        methodChannel.callFlutterEvent(
                "audioSessionStopped",
                mapOf("statusCode" to sessionStatus.statusCode?.value?.toString())
        )
        methodChannel.callFlutterMethod(MethodCall.audioSessionDidStop, null)
    }

    override fun onCameraSendAvailabilityUpdated(available: Boolean) {
        methodChannel.callFlutterEvent(
                "cameraAvailabilityChanged",
                mapOf("available" to available)
        )
    }

    override fun onConnectionBecamePoor() {
        methodChannel.callFlutterEvent("connectionBecamePoor")
    }

    override fun onConnectionRecovered() {
        methodChannel.callFlutterEvent("connectionRecovered")
    }

    override fun onRemoteVideoSourceAvailable(sources: List<RemoteVideoSource>) {
        // Out of Scope
    }

    override fun onRemoteVideoSourceUnavailable(sources: List<RemoteVideoSource>) {
        // Out of Scope
    }

    override fun onVideoSessionStarted(sessionStatus: MeetingSessionStatus) {
        methodChannel.callFlutterEvent(
                "videoSessionStarted",
                mapOf("statusCode" to sessionStatus.statusCode?.value?.toString())
        )
    }

    override fun onVideoSessionStartedConnecting() {
        methodChannel.callFlutterEvent("videoSessionConnecting")
    }

    override fun onVideoSessionStopped(sessionStatus: MeetingSessionStatus) {
        methodChannel.callFlutterEvent(
                "videoSessionStopped",
                mapOf("statusCode" to sessionStatus.statusCode?.value?.toString())
        )
    }
}

File: ios/Classes/AudioVideoObserver.swift
//
//  AudioVideoObserver.swift
//  flutter_aws_chime
//
//  Created by Conan on 2023/9/9.
//

 
import AmazonChimeSDK
import AmazonChimeSDKMedia
import Foundation

class MyAudioVideoObserver: AudioVideoObserver {
    
    var methodChannel: MethodChannelCoordinator
    
    init(withMethodChannel methodChannel: MethodChannelCoordinator) {
        self.methodChannel = methodChannel
    }
    
    func audioSessionDidStartConnecting(reconnecting: Bool) {
        methodChannel.callFlutterEvent("audioSessionConnecting", args: ["reconnecting": reconnecting])
    }

    func audioSessionDidStart(reconnecting: Bool) {
        methodChannel.callFlutterEvent("audioSessionStarted", args: ["reconnecting": reconnecting])
    }

    func audioSessionDidDrop() {
        MeetingSession.shared.meetingSession?.logger.info(msg: "Meeting session dropped")
        methodChannel.callFlutterEvent("audioSessionDropped")
    }

    func audioSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
        methodChannel.stopAudioVideoFacadeObservers()
        MeetingSession.shared.meetingSession?.logger.info(msg: "Meeting session stopped with status \(sessionStatus.description)")
        methodChannel.callFlutterEvent("audioSessionStopped", args: ["statusCode": String(sessionStatus.statusCode.rawValue)])
        methodChannel.callFlutterMethod(method: .audioSessionDidStop, args: nil)
    }

    func audioSessionDidCancelReconnect() {
        methodChannel.callFlutterEvent("audioSessionReconnectCancelled")
    }
    
    func connectionDidRecover() {
        methodChannel.callFlutterEvent("connectionRecovered")
    }

    func connectionDidBecomePoor() {
        methodChannel.callFlutterEvent("connectionBecamePoor")
    }
    
    func videoSessionDidStartConnecting() {
        MeetingSession.shared.meetingSession?.logger.info(msg: "VideoSession started connecting...")
        methodChannel.callFlutterEvent("videoSessionConnecting")
    }
    
    func videoSessionDidStartWithStatus(sessionStatus: MeetingSessionStatus) {
        MeetingSession.shared.meetingSession?.logger.info(msg:
            "VideoSession started with status \(sessionStatus.statusCode.description)")
        methodChannel.callFlutterEvent("videoSessionStarted", args: ["statusCode": String(sessionStatus.statusCode.rawValue)])
    }

    func videoSessionDidStopWithStatus(sessionStatus: MeetingSessionStatus) {
        MeetingSession.shared.meetingSession?.logger.info(msg: "VideoSession stopped with status \(sessionStatus.statusCode.description)")
        methodChannel.callFlutterEvent("videoSessionStopped", args: ["statusCode": String(sessionStatus.statusCode.rawValue)])
    }

    func remoteVideoSourcesDidBecomeAvailable(sources: [RemoteVideoSource]) {
        for remoteSourceAvailable in sources {
            MeetingSession.shared.meetingSession?.logger.info(msg: "Remote video source became available: \(remoteSourceAvailable.attendeeId)")
        }
    }
    
    func remoteVideoSourcesDidBecomeUnavailable(sources: [RemoteVideoSource]) {
        for remoteSourcesUnavailable in sources {
            MeetingSession.shared.meetingSession?.logger.info(msg: "Remote video source became available: \(remoteSourcesUnavailable.attendeeId)")
        }
    }
    
    func cameraSendAvailabilityDidChange(available: Bool) {
        methodChannel.callFlutterEvent("cameraAvailabilityChanged", args: ["available": available])
    }
}
